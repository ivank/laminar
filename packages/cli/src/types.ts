export interface Logger {
  info(args: unknown): void;
  error(args: unknown): void;
}
