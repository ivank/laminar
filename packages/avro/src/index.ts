/**
 * @packageDocumentation
 * @module @laminar/avro
 */
export { AvroDecimal, bufferToDecimal, decimalToBuffer, to64Bit } from './avro-decimal';
export { AvroTimestampMillis } from './avro-timestamp-millis';
export { AvroEpochDays } from './avro-epoch-days';
