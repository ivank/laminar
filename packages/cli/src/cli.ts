#!/usr/bin/env node

import { laminarCommand } from './commands/laminar';

laminarCommand().parse(process.argv);
