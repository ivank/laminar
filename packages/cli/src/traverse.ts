import { Document, DocumentContext } from '@laminar/ts-compose';
import { oas31 } from 'openapi3-ts';
import ts from 'typescript';

export interface AstRefMap {
  [ref: string]: unknown;
}

export interface AstContext extends DocumentContext {
  root: oas31.SchemaObject;
  refs: AstRefMap;
  convertDates?: boolean;
  optionalDefaults?: boolean;
}

export type AstConvert<TAstType = ts.TypeNode> = (
  context: AstContext,
  schema: unknown,
) => Document<TAstType, AstContext> | null;

export const isReferenceObject = (item: unknown): item is oas31.ReferenceObject =>
  typeof item === 'object' && !!item && '$ref' in item;

export const isResponseObject = (item: unknown): item is oas31.ResponseObject =>
  !isReferenceObject(item) && typeof item === 'object' && !!item && 'description' in item;

export const isMediaTypeObject = (item: unknown): item is oas31.MediaTypeObject =>
  !isReferenceObject(item) && typeof item === 'object' && !!item && 'schema' in item;

export const isRequestBodyObject = (item: unknown): item is oas31.RequestBodyObject =>
  !isReferenceObject(item) && typeof item === 'object' && !!item && 'content' in item;

export const isParameterObject = (item: unknown): item is oas31.ParameterObject =>
  !isReferenceObject(item) && typeof item === 'object' && !!item && 'in' in item && 'name' in item;

export const isSecuritySchemaObject = (item: unknown): item is oas31.SecuritySchemeObject =>
  !isReferenceObject(item) && typeof item === 'object' && !!item && 'type' in item;

export const isSchemaObject = (item: unknown): item is oas31.SchemaObject & { [index: string]: unknown } =>
  !isReferenceObject(item) && typeof item === 'object' && !!item;

export const isPathItemObject = (item: unknown): item is oas31.PathItemObject =>
  !isReferenceObject(item) && typeof item === 'object' && !!item;

export const getReferencedObject = <T>(
  item: T | oas31.ReferenceObject,
  guard: (item: unknown) => item is T,
  type: string,
  context: AstContext,
): T => {
  if (isReferenceObject(item)) {
    if (!context.refs[item.$ref]) {
      throw new Error(`Reference Error |${JSON.stringify(item)}| not found`);
    }
    const resolved = getReferencedObject(context.refs[item.$ref], guard, type, context);
    if (!guard(resolved)) {
      throw new Error(
        `Reference Error, the reference |${JSON.stringify(item)}| which was resolved to |${JSON.stringify(
          resolved,
        )}| was not the expected type "${type}"`,
      );
    }
    return resolved;
  } else {
    return item;
  }
};
