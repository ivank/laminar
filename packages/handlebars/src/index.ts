/**
 * @packageDocumentation
 * @module @laminar/handlebars
 */

import { Middleware } from '@laminar/laminar';
import { readdirSync, readFileSync, existsSync } from 'fs';
import { compile, RuntimeOptions, TemplateDelegate } from 'handlebars';
import { join, relative } from 'path';
import { statSync } from 'fs';

export type Cache = Map<string, { template: TemplateDelegate; mtime: Date }>;

/**
 * Cache types.
 * - preload: load all the templates ones into memory
 * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
 * - none: do not cache templates and load them on every request.
 */
export type CacheType = 'preload' | 'expiry' | 'none';

export interface TemplateConfig {
  /**
   * Helper functions passed to handlebar templates
   */
  helpers?: RuntimeOptions['helpers'];
  /**
   * Main directory where handlebar files should be located, defaults to current working directory.
   * Looks for "partials" and "views" directories inside of it.
   */
  dir?: string;
  /**
   * Directory where all the partials are located, inside the main 'dir' directory, defaults to 'partials'
   */
  partials?: string;
  /**
   * Name of the handlebars file extension. Defaults to 'handlebars', so that only 'my-file.handlebars' kind of files are loaded.
   */
  extension?: string;
  /**
   * Directory name where all the views are located, inside the main 'dir' directory, defaults to 'views'
   */
  views?: string;
  /**
   * Cache types.
   * - preload: load all the templates ones into memory
   * - expiry: load partials when needed and keep them in cache, but check the file's mtime and reload template if changed
   * - none: do not cache templates and load them on every request.
   *
   * Defaults to 'preload'
   */
  cacheType?: CacheType;
  /**
   * Global data to be passed to every template
   */
  globalData?: Record<string, unknown>;
}

export interface CompileTemplatesOptions {
  dir: string;
  extension: string;
}

export type HandlebarsRender = (view: string, data?: Record<string, unknown>) => string;

export interface HandlebarsContext {
  hbs: HandlebarsRender;
}

const hasExtension =
  (extension: string) =>
  (file: string): boolean =>
    file.endsWith(extension);

export const deepReaddirSync = ({ dir, extension }: CompileTemplatesOptions): string[] =>
  readdirSync(dir, { withFileTypes: true })
    .reduce<string[]>((all, item) => {
      return all.concat(
        item.isDirectory() ? deepReaddirSync({ dir: join(dir, item.name), extension }) : join(dir, item.name),
      );
    }, [])
    .filter(hasExtension(extension));

const toName = (dir: string, file: string): string => relative(dir, file.slice(0, file.lastIndexOf('.')));

export const compileTemplates = ({ dir, extension }: CompileTemplatesOptions): Array<[string, TemplateDelegate]> =>
  deepReaddirSync({ dir, extension }).reduce<Array<[string, TemplateDelegate]>>(
    (all, file) => [...all, [toName(dir, file), compile(readFileSync(file, 'utf8'))]],
    [],
  );

export const loadTemplate = (
  name: string,
  cache: Cache,
  cacheType: CacheType,
  { dir, extension }: CompileTemplatesOptions,
): TemplateDelegate => {
  const file = `${join(dir, name)}.${extension}`;

  const stat = cacheType === 'expiry' ? statSync(file) : { mtime: new Date(0) };
  const cached = cache.get(name);

  if (cacheType !== 'none' && cached && stat.mtime < cached.mtime) {
    return cached.template;
  } else {
    const template = compile(readFileSync(file, 'utf8'));
    cache.set(name, { template, mtime: cacheType === 'expiry' ? stat.mtime : new Date() });
    return template;
  }
};

const loadPartials = (options: CompileTemplatesOptions): { [key: string]: TemplateDelegate } =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce((acc, [name, template]) => ({ ...acc, [name]: template }), {})
    : {};

const loadTemplates = (cache: Cache, options: CompileTemplatesOptions): Cache =>
  existsSync(options.dir)
    ? compileTemplates(options).reduce(
        (cache, [name, template]) => cache.set(name, { template, mtime: new Date() }),
        cache,
      )
    : cache;

export const handlebars = ({
  helpers = {},
  dir = process.cwd(),
  extension = 'handlebars',
  partials = 'partials',
  views = 'views',
  cacheType = 'preload',
  globalData = {},
}: TemplateConfig): HandlebarsRender => {
  const cache = cacheType === 'preload' ? loadTemplates(new Map(), { dir, extension }) : new Map();
  const partialTemplates = cacheType === 'preload' ? loadPartials({ dir: join(dir, partials), extension }) : undefined;

  return (name: string, data: Record<string, unknown> = {}) => {
    const template = loadTemplate(name, cache, cacheType, { dir: join(dir, views), extension });
    return template(
      { ...globalData, ...data },
      { helpers, partials: partialTemplates ?? loadPartials({ dir: join(dir, partials), extension }) },
    );
  };
};

export const handlebarsMiddleware = (config: TemplateConfig): Middleware<HandlebarsContext> => {
  const hbs = handlebars(config);
  return (next) => (req) => next({ ...req, hbs });
};
