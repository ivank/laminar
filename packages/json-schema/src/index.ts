/**
 * @packageDocumentation
 * @module @laminar/json-schema
 */

export { ValidationError } from './ValidationError';
export { ResolveError } from './ResolveError';
export {
  InvalidCode,
  Invalid,
  Messages,
  Options,
  Validator,
  Validation,
  error,
  empty,
  onlyErrors,
  errors,
  evaluateItem,
  evaluateProperty,
  skipRest,
  hasErrors,
  hasEvaluatedItem,
  hasEvaluatedProperty,
  combine,
  combineErrors,
  childOptions,
  validateSchema,
} from './validation';
export { Schema, JsonSchema, PrimitiveType } from './schema';
export { toMessages, toMessage, messages } from './messages';
export {
  isCompiled,
  withinContext,
  toSchemaObject,
  compile,
  ResolvedSchema,
  resolve,
  resolveFile,
  RefMap,
  FileContext,
} from './resolve';
export {
  validate,
  validateCompiled,
  ensureValid,
  ensureValidCompiled,
  ValidateOptions,
  ValidateSchemaOptions,
  ValidateCompiledOptions,
  Result,
  ResultError,
  ResultSuccess,
  Drafts,
  FormatErrors,
} from './validate';
export { CoerceCompiledOptions, CoerceSchemaOptions, CoerceOptions, coerce, coerceCompiled } from './coerce';
export { coerceSchema, coercers, CoercerOptions, Coercer } from './coercion';
