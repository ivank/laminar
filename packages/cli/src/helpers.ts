import { compile, ensureValid, Schema } from '@laminarjs/json-schema';
import { openapiV3, openapiV31 } from 'openapi-schemas';
import { oas31 } from 'openapi3-ts';
import { AstContext } from './traverse';
import * as YAML from 'yaml';

export const concatStreamToString = async (stream: NodeJS.ReadStream): Promise<string> => {
  let str = '';

  return new Promise((resolve, reject) => {
    stream.on('data', (data) => (str += data.toString()));
    stream.on('end', () => resolve(str));
    stream.on('error', (error) => reject(error));
  });
};

export const toTitleCase = (str: string): string =>
  str
    .replace('_', ' ')
    .replace(/\w\S*/g, (txt) => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase())
    .replace(' ', '');

const escapeRegExp = (value: string) => value.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

const matchMediaQuery = <T>(mediaType: string, resources: Record<string, T>): T | undefined => {
  const [base, type] = mediaType.split('/');
  const matcher = new RegExp(`^(${base}/\\w+\\+${escapeRegExp(type)})|${base}/\\*|\\*/\\*$`);
  const key = Object.keys(resources).find((key) => matcher.test(key));
  return key ? resources[key] : undefined;
};

export const findMediaType = <T>(mediaType: string, resources?: Record<string, T>): T | undefined =>
  resources ? resources[mediaType] ?? matchMediaQuery(mediaType, resources) : undefined;

/**
 * Compile an openapi schema, converting it into a context.
 * This context allows you to use it in conjuction with parts of the schema as
 * all the urls have been resolved and the validation can be done statically
 */
export const toContext = async (
  fileName: string,
): Promise<{ context: AstContext; uris: string[]; value: oas31.OpenAPIObject }> => {
  const { schema, uris, refs } = await compile(fileName);

  const openapiSchema =
    typeof schema === 'object' &&
    'openapi' in schema &&
    typeof schema.openapi === 'string' &&
    schema.openapi.startsWith('3.1')
      ? openapiV31
      : openapiV3;

  const { value } = await ensureValid<oas31.OpenAPIObject>({
    schema: openapiSchema as Schema,
    value: schema,
    name: 'OpenAPI',
  });

  return { context: { root: value, refs }, value, uris };
};

/**
 * Parse a string into a Schema object.
 * Supports 'json' and 'yaml'
 */
export const parseSchema = (type: 'json' | 'yaml', content: string): Schema => {
  switch (type) {
    case 'json':
      return JSON.parse(content);
    case 'yaml':
      return YAML.parse(content);
  }
};

/**
 * regex for matching ANSI escape codes.
 *
 * Example:
 *
 * ```js
 *  '\u001B[4mcake\u001B[0m'.match(ansiRegex); => ['\u001B[4m', '\u001B[0m']
 * ```
 * Source: https://github.com/chalk/ansi-regex#readme
 */
export const ansiRegex = new RegExp(
  [
    '[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)',
    '(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-ntqry=><~]))',
  ].join('|'),
  'g',
);
