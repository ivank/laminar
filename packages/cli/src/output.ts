import chalk from 'chalk';
import { ansiRegex } from './helpers';

export const ansiLength = (text: string): number => text.replace(ansiRegex, '').length;

export const table = (rows: string[][]): string => {
  const columns = rows[0].map((_, index) => rows.map((row) => row[index]));
  const columnLengths = columns.map((column) => column.reduce((len, item) => Math.max(len, ansiLength(item)), 0));

  return rows
    .map((row) =>
      row.map((item, index) => item + ' '.repeat(columnLengths[index] - ansiLength(item))).join(chalk.gray(' | ')),
    )
    .join('\n');
};
