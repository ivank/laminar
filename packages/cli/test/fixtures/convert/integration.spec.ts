import { fixturesConvert } from '../../../src';

const tables = [
  {
    name: 'test',
    columns: [
      { name: 'id', type: 'integer', isOptional: false },
      { name: 'name', type: 'varchar', isOptional: true },
      { name: 'title', type: 'test_title', isOptional: false },
      { name: 'created_at', type: 'timestamp', isOptional: false },
      { name: 'updated_at', type: 'timestamp without time zone', isOptional: true },
      { name: 'end_on', type: 'date', isOptional: true },
      { name: 'is_active', type: 'boolean', isOptional: false },
    ],
  },
];

const enums = [{ name: 'test_title', enum: ['mr', 'mrs'] }];

describe('Fixtures ts test', () => {
  it('Should convert fixtures successfully', () => {
    expect(fixturesConvert({ tables, enums })).toMatchSnapshot('Normal');
    expect(fixturesConvert({ tables, enums, suffix: 'Fixture' })).toMatchSnapshot('Suffix');
    expect(fixturesConvert({ tables, enums, suffix: 'Fixture', titleCase: true })).toMatchSnapshot('TitleCase');
  });
});
