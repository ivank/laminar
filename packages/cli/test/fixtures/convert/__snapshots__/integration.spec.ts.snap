// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fixtures ts test Should convert fixtures successfully: Normal 1`] = `
"import { Fixture } from "@laminar/fixtures";

/**
 * Fixtures generated for postgres tables */

/**
 * Fixture type for table "test" with column types loaded from the database
 * You can use the type with a fixutre, or by creating a fixture builder
 *
 * \`\`\`typescript
 * import { fixture, BuildFixture } from '@laminar/fixtures';
 * const mytest = fixture<test>({
 *   // ...
 * });
 *
 * const buildtest: BuildFixture<test> => ({ columns }) => fixture({
 *   // ...
 *   ...columns
 * });
 * \`\`\`
 */
export type test = Fixture<{
    id: number;
    name?: string;
    title: "mr" | "mrs";
    created_at: Date;
    updated_at?: Date;
    end_on?: Date;
    is_active: boolean;
}>;
"
`;

exports[`Fixtures ts test Should convert fixtures successfully: Suffix 1`] = `
"import { Fixture } from "@laminar/fixtures";

/**
 * Fixtures generated for postgres tables */

/**
 * Fixture type for table "test" with column types loaded from the database
 * You can use the type with a fixutre, or by creating a fixture builder
 *
 * \`\`\`typescript
 * import { fixture, BuildFixture } from '@laminar/fixtures';
 * const mytestFixture = fixture<testFixture>({
 *   // ...
 * });
 *
 * const buildtestFixture: BuildFixture<testFixture> => ({ columns }) => fixture({
 *   // ...
 *   ...columns
 * });
 * \`\`\`
 */
export type testFixture = Fixture<{
    id: number;
    name?: string;
    title: "mr" | "mrs";
    created_at: Date;
    updated_at?: Date;
    end_on?: Date;
    is_active: boolean;
}>;
"
`;

exports[`Fixtures ts test Should convert fixtures successfully: TitleCase 1`] = `
"import { Fixture } from "@laminar/fixtures";

/**
 * Fixtures generated for postgres tables */

/**
 * Fixture type for table "test" with column types loaded from the database
 * You can use the type with a fixutre, or by creating a fixture builder
 *
 * \`\`\`typescript
 * import { fixture, BuildFixture } from '@laminar/fixtures';
 * const myTestFixture = fixture<TestFixture>({
 *   // ...
 * });
 *
 * const buildTestFixture: BuildFixture<TestFixture> => ({ columns }) => fixture({
 *   // ...
 *   ...columns
 * });
 * \`\`\`
 */
export type TestFixture = Fixture<{
    id: number;
    name?: string;
    title: "mr" | "mrs";
    created_at: Date;
    updated_at?: Date;
    end_on?: Date;
    is_active: boolean;
}>;
"
`;
